# How to prepare input bam for Kourami

## Input to Kourami
The input to Kourami is a bam file containing read alignments of a subset of reads from HLA loci to known HLA reference panel generated by incorporating gene-length MSA and exon-only MSA provided by IMGT/HLA database. Currently, Kourami uses 3.24.0 release of the database. Different versions of the database can be used easily and the document explaining how to swap the db will be updated shortly. 

## What you need 
0. [Kourami]
1. [bamUtil](https://github.com/statgen/bamUtil/releases) 1.0.13 or higher
2. [samtools](https://github.com/samtools/samtools) 1.3.1 or higher
3. [bwa 0.7.15-r1140](https://github.com/lh3/bwa) or higher
4. [bwa-kit 0.7.12](https://github.com/lh3/bwa/tree/master/bwakit) if you don't already have GRCh38 alignment


## Download a suitable flavor of GRCh38
The current version (GRCh38) of the human genome comes in multiple flavors because they are published as multiple components. The components are:

&#9398;. Primary assembly : chromosome, unplaced, and unlocaized contigs + EBV (195 contigs)

&#9399;. Decoy (2386 contigs) 

&#9400;. ALT : ALT haplotype (261 contigs)

&#9401;. HLA alleles packaged in hs38DH in [bwa.kit](https://github.com/lh3/bwa/tree/master/bwakit) (525 contigs)

We find that either using hs38NoAltDH ( a + b + d ) or hs38DH ( a + b + c + d ) is most effective for extracting reads from HLA loci. The hs38DH flavor is used by 1000 Genome project (see [here](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/README.1000genomes.GRCh38DH.alignment))

You need to download hs38NoAltDH by running download_grch38.sh script located under `script` directory
```
kourami@kourami:~/kourami$./scripts/download_grch38.sh hs38NoAltDH
```
This will generate the reference fasta file `hs38NoAltDH.fa` under `resources` directory. Then you need to `bwa index` the downloaded reference by running:
```
kourami@kourami:~/kourami$bwa index resources/hs38NoAltDH.fa	
```

## Read Extraction and input bam generation from GRCh38 bam
If you have WGS data aligned to GRCh38 reference, we first need to extract reads that likely coming from HLA loci. If not, see the section "When aligned bam files to GRCh38 are not available" below. Depending on the GRCh38 flavor your bam is aligned to, you need to use the correct script to extract reads (See Table below)

GRCh38 flavor | Use
------------------|--------------
hs38DH | alignAndExtract_hs38DH.sh
hs38NoAltDH | alignAndExtract_hs38DH_NoAlt.sh
&#9398; + (&#9399; optional) + &#9400; | alignAndExtract_hs38Alt.sh
&#9398; + (&#9399; optional) [NOT recommended] | alignAndExtract_hs38.sh

#### Running the extraction (pre-processing) script
An example is shown for bam aligned to hs38DH below:
````
kourami@kourami:~/kourami$mkdir test
kourami@kourami:~/kourami$cd test
kourami@kourami:~/kourami/test$./alignAndExtract_hs38DH.sh NA12878 /mnt/data/NA12878.hs38DH.bam
````
This will generate `NA12878_on_KouramiPanel.bam` and this can be fed into Kourami.

## When aligned bam files to GRCh38 are not available:

When an aligned bam file (to the human genome) is not available, you must first align high coverage WGS data ( >30X coverage ) to the reference human genome, we recommend using the hs38NoAltDH or hs38DH flavor (see "Downloading the correct version of GRCh38" section above). We recommend you to follow 1000 genomes GRCh38 pipeline explained [here](http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/README.1000genomes.GRCh38DH.alignment). 
